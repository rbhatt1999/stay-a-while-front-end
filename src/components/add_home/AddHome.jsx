import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import style from './AddHome.module.css';
import Images from './Images';
import { addHome } from './addHomeSlice';

const AddHome = () => {
  const navigate = useNavigate();

  const formInputState = {
    name: '',
    location: '',
    description: '',
    price: '',
    no_of_rooms: '',
    rating: '1',
    images: [''],
  };
  const [formInput, setformInput] = useState(formInputState);
  const dispatch = useDispatch();

  const handleChange = (e) => {
    const { name, value } = e.target;
    setformInput({ ...formInput, [name]: value });
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const newHome = {
      home_stay: {
        name: formInput.name,
        location: formInput.location,
        description: formInput.description,
        price: formInput.price,
        no_of_rooms: formInput.no_of_rooms,
        rating: formInput.rating,
      },
      images: formInput.images,
    };

    dispatch(addHome(newHome));
    setformInput(formInputState);
    navigate('/');
  };

  const handleAdd = () => {
    setformInput({ ...formInput, images: [...formInput.images, ''] });
  };

  const handleRemove = (index) => {
    const images = [...formInput.images];
    images.splice(index, 1);
    setformInput({ ...formInput, images });
  };

  const handleImageChange = (e, index) => {
    const images = [...formInput.images];
    images[index] = e.target.value;
    setformInput({ ...formInput, images });
  };

  return (
    <section id={style.add_home}>
      <form onSubmit={handleSubmit} className={style.form_container}>
        <input
          type="text"
          name="name"
          value={formInput.name}
          onChange={handleChange}
          required
          placeholder="Name of the property"
          minLength={3}
        />
        <input
          type="text"
          name="location"
          value={formInput.location}
          onChange={handleChange}
          required
          placeholder="Location of the property"
          minLength={3}
        />
        <textarea
          type="textarea"
          name="description"
          value={formInput.description}
          onChange={handleChange}
          required
          placeholder="Description of the property"
          minLength={3}
          maxLength={1000}
          rows="7"
          cols="60"
        />
        <input
          type="number"
          name="price"
          value={formInput.price}
          onChange={handleChange}
          required
          placeholder="Price per day"
          min={1}
        />
        <input
          type="number"
          name="no_of_rooms"
          value={formInput.no_of_rooms}
          onChange={handleChange}
          required
          placeholder="Number of rooms"
          min={1}
        />
        <select
          name="rating"
          value={formInput.rating}
          onChange={handleChange}
          required
        >
          <option value="1">1 star</option>
          <option value="2">2 stars</option>
          <option value="3">3 stars</option>
          <option value="4">4 stars</option>
          <option value="5">5 stars</option>
        </select>
        <Images
          form={formInput}
          onAdd={handleAdd}
          onRemove={handleRemove}
          onChange={handleImageChange}
        />
        <input type="submit" value="Submit" className={style.submission} />
      </form>
    </section>
  );
};

export default AddHome;
