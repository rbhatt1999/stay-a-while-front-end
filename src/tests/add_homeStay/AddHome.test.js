import React from 'react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import {
  render, act, screen,
} from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import addHomeSlice from '../../components/add_home/addHomeSlice';
import AddHome from '../../components/add_home/AddHome';

jest.mock('axios');

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));

const store = configureStore({
  reducer: {
    addHome: addHomeSlice,
  },
});

describe('AddHome', () => {
  beforeEach((async () => {
    await act(async () => {
      render(
        <Provider store={store}>
          <AddHome />
        </Provider>,
      );
    });
  }));

  test('should have a submit button', () => {
    expect(screen.getByRole('button', { name: /submit/i })).toBeInTheDocument();
  });

  test('should have a input field with placeholdeer "Name of the property"', () => {
    expect(screen.getByPlaceholderText(/Name of the property/i)).toBeInTheDocument();
  });

  test('should have a input field with placeholdeer "Location"', () => {
    expect(screen.getByPlaceholderText(/Location of the property/i)).toBeInTheDocument();
  });

  test('should have a input field with placeholdeer "City"', () => {
    expect(screen.getByPlaceholderText(/Description of the property/i)).toBeInTheDocument();
  });

  test('should have a input field with placeholdeer "State"', () => {
    expect(screen.getByPlaceholderText(/Price per day/i)).toBeInTheDocument();
  });

  test('should have a input field with placeholdeer "Zip"', () => {
    expect(screen.getByPlaceholderText(/Number of rooms/i)).toBeInTheDocument();
  });
});

describe('AddHome', () => {
  test('should match snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <AddHome />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
