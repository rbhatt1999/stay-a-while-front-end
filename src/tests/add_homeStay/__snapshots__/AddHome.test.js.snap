// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddHome should match snapshot 1`] = `
<section
  id="add_home"
>
  <form
    className="form_container"
    onSubmit={[Function]}
  >
    <input
      minLength={3}
      name="name"
      onChange={[Function]}
      placeholder="Name of the property"
      required={true}
      type="text"
      value=""
    />
    <input
      minLength={3}
      name="location"
      onChange={[Function]}
      placeholder="Location of the property"
      required={true}
      type="text"
      value=""
    />
    <textarea
      cols="60"
      maxLength={1000}
      minLength={3}
      name="description"
      onChange={[Function]}
      placeholder="Description of the property"
      required={true}
      rows="7"
      type="textarea"
      value=""
    />
    <input
      min={1}
      name="price"
      onChange={[Function]}
      placeholder="Price per day"
      required={true}
      type="number"
      value=""
    />
    <input
      min={1}
      name="no_of_rooms"
      onChange={[Function]}
      placeholder="Number of rooms"
      required={true}
      type="number"
      value=""
    />
    <select
      name="rating"
      onChange={[Function]}
      required={true}
      value="1"
    >
      <option
        value="1"
      >
        1 star
      </option>
      <option
        value="2"
      >
        2 stars
      </option>
      <option
        value="3"
      >
        3 stars
      </option>
      <option
        value="4"
      >
        4 stars
      </option>
      <option
        value="5"
      >
        5 stars
      </option>
    </select>
    <div
      className="imageContainer"
    >
      <div
        className="imageHeading"
      >
        <h5>
          Add images of the property
          <button
            className="imageButton"
            onClick={[Function]}
            type="button"
          >
             
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16 8A8 8 0 1 1 0 8a8 8 0 0 1 16 0zM8.5 4.5a.5.5 0 0 0-1 0v3h-3a.5.5 0 0 0 0 1h3v3a.5.5 0 0 0 1 0v-3h3a.5.5 0 0 0 0-1h-3v-3z"
              />
            </svg>
          </button>
        </h5>
      </div>
      <div
        className="image_container"
      >
        <input
          name="image"
          onChange={[Function]}
          placeholder="Image URL"
          required={true}
          type="text"
          value=""
        />
        <button
          onClick={[Function]}
          type="button"
        >
           
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M864 256H736v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zm-200 0H360v-72h304v72z"
            />
          </svg>
        </button>
      </div>
    </div>
    <input
      className="submission"
      type="submit"
      value="Submit"
    />
  </form>
</section>
`;
